vmas/simulator/environment/environment.py

```diff
@@ -503,41 +504,48 @@ def _set_action(self, action, agent):
             if not self.multidiscrete_actions:
                 # This bit of code translates the discrete action (taken from a space that
                 # is the cartesian product of all action spaces) into a multi discrete action.
-                # For example, if agent.action_size=4, it will mean that the agent will have
-                # 4 actions each with 3 possibilities (stay, decrement, increment).
-                # The env will have a space Discrete(3**4).
-                # This code will translate the action (with shape [n_envs,1] and range [0,3**4)) to an
-                # action with shape [n_envs,4] and range [0,3).
-                n_actions = self.get_agent_action_space(agent).n
-                action_range = torch.arange(n_actions, device=self.device).expand(
-                    self.world.batch_dim, n_actions
-                )
-                physical_action = action
-                action_range = torch.where(action_range == physical_action, 1.0, 0.0)
-                action_range = action_range.view(
-                    (self.world.batch_dim,)
-                    + (3,) * agent.action_size
-                    + (self.world.dim_c,)
-                    * (1 if not agent.silent and self.world.dim_c != 0 else 0)
+                # This is done by iteratively taking the modulo of the action and dividing by the
+                # number of actions in the current action space, which treats the action as if
+                # it was the "flat index" of the multi-discrete actions. E.g. if we have
+                # nvec = [3,2], action 0 corresponds to the actions [0,0],
+                # action 1 corresponds to the action [0,1], action 2 corresponds
+                # to the action [1,0], action 3 corresponds to the action [1,1], etc.
+                flat_action = action.squeeze(-1)
+                actions = []
+                nvec = list(agent.discrete_action_nvec) + (
+                    [self.world.dim_c]
+                    if not agent.silent and self.world.dim_c != 0
+                    else []
                 )
-                action = action_range.nonzero()[:, 1:]
+                for n in nvec:
+                    actions.append(flat_action % n)
+                    flat_action = flat_action // n
+                action = torch.stack(actions, dim=-1)
```

Comment:
There also seems to be a little inconsistency with this bit

with `nvec=[3,3]` and `action=tensor([7, 6, 3, 0, 1, 7, 5, 6, 3, 7, 4, 4, 7, 7, 1, 3, 2, 3, 6, 3, 0, 6, 8, 0,
        3, 2, 2, 3, 4, 5, 8, 8])`

i am getting

## PR

```
tensor([[1, 2],
        [0, 2],
        [0, 1],
        [0, 0],
        [1, 0],
        [1, 2],
        [2, 1],
        [0, 2],
        [0, 1],
        [1, 2],
        [1, 1],
        [1, 1],
        [1, 2],
        [1, 2],
        [1, 0],
        [0, 1],
        [2, 0],
        [0, 1],
        [0, 2],
        [0, 1],
        [0, 0],
        [0, 2],
        [2, 2],
        [0, 0],
        [0, 1],
        [2, 0],
        [2, 0],
        [0, 1],
        [1, 1],
        [2, 1],
        [2, 2],
        [2, 2]])
```

## vmas main

```
tensor([[2, 1],
        [2, 0],
        [1, 0],
        [0, 0],
        [0, 1],
        [2, 1],
        [1, 2],
        [2, 0],
        [1, 0],
        [2, 1],
        [1, 1],
        [1, 1],
        [2, 1],
        [2, 1],
        [0, 1],
        [1, 0],
        [0, 2],
        [1, 0],
        [2, 0],
        [1, 0],
        [0, 0],
        [2, 0],
        [2, 2],
        [0, 0],
        [1, 0],
        [0, 2],
        [0, 2],
        [1, 0],
        [1, 1],
        [1, 2],
        [2, 2],
        [2, 2]])

```

Reply:
`action = torch.stack(actions, dim=-1).flip(-1)` this should fix it

Reply:
Thanks for catching this! I fixed it and added relevant tests